server/src/test/java/org/cloudfoundry/identity/uaa/user/UserInfoTest.java

```diff
@@ -3,70 +3,90 @@
 
 class UserInfoTest {
 
     @Nested
     class Equals {
 
         @Test
+        @SuppressWarnings("java:S5863")
         void expectTrueWhenBothUserInfoObjectsAreSame() {
             UserInfo userInfo = new UserInfo();
-            assertTrue(userInfo.equals(userInfo));
+            assertEquals(userInfo, userInfo);
         }
 
         @Test
         void expectFalseWhenUserInfoIsComparedWithOtherObject() {
             UserInfo userInfo = new UserInfo();
-            assertFalse(userInfo.equals(new Object()));
+            assertNotEquals(userInfo, new Object());
         }
 
         @Test
         void expectTrueWhenBothUserInfoWithoutAnyRole() {
             UserInfo u1 = new UserInfo();
             UserInfo u2 = new UserInfo();
-            assertTrue(u1.equals(u2));
+            assertEquals(u1, u2);
         }
 
         @Test
         void expectFalseWhenOnlyOneUserInfoHasRole() {
             UserInfo u1 = userInfoWithRoles(List.of("group1"));
             UserInfo u2 = new UserInfo();
-            assertFalse(u1.equals(u2));
+            assertNotEquals(u1, u2);
         }
 
         @Test
         void expectTrueWhenBothUserInfoHaveSameRole() {
             UserInfo u1 = userInfoWithRoles(List.of("group1"));
             UserInfo u2 = userInfoWithRoles(List.of("group1"));
-            assertTrue(u1.equals(u2));
+            assertEquals(u1, u2);
         }
 
         @Test
         void expectFalseWhenBothUserInfoHaveDifferentRoles() {
             UserInfo u1 = userInfoWithRoles(List.of("group1"));
             UserInfo u2 = userInfoWithRoles(List.of("group2"));
-            assertFalse(u1.equals(u2));
+            assertNotEquals(u1, u2);
         }
 
         @Test
         void expectFalseWhenBothUserInfoHaveMultipleRolesAndFewAreCommon() {
             UserInfo u1 = userInfoWithRoles(List.of("group1", "group2", "group3"));
             UserInfo u2 = userInfoWithRoles(List.of("group2", "group3", "group4"));
-            assertFalse(u1.equals(u2));
+            assertNotEquals(u1, u2);
         }
 
         @Test
         void expectTrueWhenBothUserInfoHaveMultipleRolesAndAllAreCommon() {
             UserInfo u1 = userInfoWithRoles(List.of("group1", "group2", "group3"));
             UserInfo u2 = userInfoWithRoles(List.of("group1", "group2", "group3"));
-            assertTrue(u1.equals(u2));
+            assertEquals(u1, u2);
         }
 
+        @Test
+        void expectTrueWhenBothUserInfoHaveManyMultipleRolesAndAllAreCommon() {
+            // Given
+            int count = 10000;
+            List<String> roleList1 = new ArrayList<>(count);
+            List<String> roleList2 = new ArrayList<>(count);
+            for (int i=1; i <= count; i++ ) {
+                roleList1.add("groups".concat(Integer.toString(i)));
+            }
+            for (int i=count; 0 < i; i-- ) {
+                roleList2.add("groups".concat(Integer.toString(i)));
+            }
```

Comment:
Did you consider adding some duplicates in the list, to validate that the code deals with that properly?